### ISSUE TITLE
BIOS emulated PS2 keyboard does not work

### ISSUE THREAD

**Issue by jm--** [jm--/seL4Doom#1]
Finally, I found the time to try it on real hardware. I boot from an USB stick, and I can get to the main screen. :) But the keyboard does not work at all. :(

By now I have some idea of what's going on: I use a laptop with an external USB keyboard (because the built-in keyboard died on me), and the BIOS is configured to emulate legacy PS2 devices. I wrote a little test program, and I read up a bit on keyboard scan codes. It seems the problem is that my keyboard generates "Scan Code Set 1" key codes, which the platsupport keyboard code was not designed to handle, I think. See here for info about scan codes: http://wiki.osdev.org/PS2_Keyboard#Scan_Code_Sets  

The good news is that my seL4 test program can read all keys (including shift, alt, cursor)! So it should be possible to hack something together to make it work!

**Comment by jm--** [comment]
Okay, so I spent some time reviewing keyboard_ps2.c in libplatsupport. It seems the keyboard (controller) of my laptop has at least two problems with it, but I blame my (buggy) BIOS, not libplatsupport.

1) The self test code resets the already set command register. (QEMU and VirtualBox do not do that.) That's bad because it results in the keyboard operating in scan code set 1.

```
ps2_single_control(&state->ops, PS2_CMD_CONTROLLER_SELF_TEST);
uint8_t res = ps2_read_output(&state->ops);
if (res != PS2_CONTROLLER_SELF_TEST_OK) {
    return -1;
}
```

In a test program, I fixed this by moving the self test code to the beginning of the init-keyboard routine where it does not matter that it overwrites an uninitialized value.

2) The code to initialize the command register (to enable IRQ and disable translation) has some serious race conditions. The value is only really set when I introduce some delays. I have to resort to something ugly as this to make it work: Set and re-read till value is really set.

```
int i = 0;
int8_t config;
int8_t config2;

/* Enable IRQs and disable translation (IRQ bits 0, 1, translation 6). */
ps2_single_control(&state->ops, PS2_READ_CMD_BYTE);
config = ps2_read_output(&state->ops);
config |= 0x1;
config &= 0xBF;
do {
    //Write 8042 Command Byte: next data byte written to port 60h is
    //          placed in 8042 command register.
    ps2_single_control(&state->ops, PS2_WRITE_CMD_BYTE);
    mywait(i * 1000);
    ps2_write_output(&state->ops, config);

    //increase delay
    mywait(i * 1000);   // does some math in a loop
    ps2_single_control(&state->ops, PS2_READ_CMD_BYTE);
    config2 = ps2_read_output(&state->ops);
} while (config2 != config && i++ < 100);
```

Am I missing something? Is there a better way?

And there are other minor issues: e.g. disabling the keyboard interface does not always set the appropriate bit in the command register.

Anyway, with the two "fixes" in place, I can essentially always boot my laptop with the built-in keyboard operating in scan code set 2, which libplatsupport expects. Nice! Unfortunately, my external USB keyboard still operates in scan code set 1, but at this point I cannot think of anything else that could be done to activate scan code set 2 because the command to return the currently active scan-code set already returns 2. Hmmm...

So the only option may be to support _both_ scan codes (1 and 2) somehow...

**Comment by jm--** [comment]
Part A) Make the keyboard of my laptop work:
I've imported the PS/2 keyboard code from libplatsupport and implemented the two changes described above (see 6482e6965747d423e2c10167541e3842e754283c).

Part B) Make my external USB keyboard work:
I have not yet found a way to set my external USB keyboard to use scan code set 2, so the problem is that (with the above fix) my built-in keyboard is using scan code set 2 but my external keyboard scan code set 1.

My current "solution" is that the keyboard handling code uses either scan code set 1 _or_ 2 depending on the scan code of the first key pressed. That way I have an easy and dynamic method to determine which keyboard I want to use. (I currently cannot use both at the same time.) (See fd654c82a3e3d9673b921ca3efaea9aaff8c61aa, 122315b9ee711270df751f2a5671a02aa540f0f4)

Unfortunately, there is yet another new problem. When using the USB keyboard, I'm experiencing some keyboard lag. For example, releasing the right arrow does not take effect right away: I keep turning to the right for some time. I believe this is because the keyboard's key repeat rate is too high. I wrote
a function to set the repeat rate (c9713bde204a111e4f6c973ff29ac9c61457722d), but it works only with the built-in keyboard, not the external one. (So it's the same behavior as seen with setting the scan code set.) I think I don't want to add yet another hack to make this one work, at least not now, for the built-in keyboard works (aside the keys that are genuinely broken).

I now can play DOOM on real hardware, on my laptop: Victory at last!
